### DISCUSSION TITLE
Custom contact model in PyBullet?

### DISCUSSION THREAD

**Discussion by Mark799** [bulletphysics/bullet3#4057] [Ideas]
Hello,

I'm a graduate student at Delft University of Technology, and I'm researching the possibilities of PyBullet for modelling viscous materials. Thus far, me and my supervisors are very positive and enthusiastic about the software. 

To further enhance the software for this purpose, I would like to propose and add-on to the library. Namely, being able to set a custom contact force between elements, based on contact distance and relative velocity at the location of contact. 

Currently, I've tried to set the stiffness and damping of all objects to zero, and apply an external force instead. This works reasonable, but requires small time stepping, as I can't utilize the iterations per time step. 

Therefore, I hope that this suggestion can be implemented!

Kind regards,
Mark

UPDATE: Setting specific dynamic properties, like stiffness and damping between two objects would also be of great help. So, for example: additional to _changeDynamics_(bodyUniqueId, ...), have a function like _changeDynamics2_(bodyUniqueIdA, bodyUniqueIdB, ...)

**Comment by lasagnaphil** [comment]
I've studied a bit myself on how contact resolution works in PyBullet, and I think it will be hard to integrate unless there are big changes to the engine code. Right now Bullet simplifies the contact/friction model a lot so that it can be solved by conventional LCP solvers in real-time speeds. Using a more sophisticated contact model (for example, like anisotrophic friction) would either require using some tricks to convert this into an LCP constraint, or ditching the LCP formulation altogether and rewriting the constraint solver.
